### PR TITLE
Activity special substitutions round 2

### PR DESCRIPTION
**Pull request by yufengwng**
## Context

A follow-up to DDP-4703.

This PR proposes an alternative design for "activity special substitutions" (e.g. substituting in participant's name into activity). Previously, we're plugging in constant string values. In new design, we provide a `ddp` object which you can call methods on within the template to get certain values to substitute with. Advantages over current design:

* Minimal variable name collision. Only variable name you can't use is `ddp`, whereas current design reserves a few other names.
* More flexible since methods can take arguments. This means we don't have to add different special variables to get dates in different formats -- now it's just one method to call.
* Should be easier to extend. New values can be provided by adding new methods to `RenderValueProvider`.

This is possible with the built-in support of [Apache Velocity](https://velocity.apache.org/engine/1.7/) library that we're using.

## Checklist

- [x] I have labeled the type of changes involved using the `C-*` labels.
- [x] I have assessed potential risks and labeled using the `R-*` labels.
- [x] I have considered error handling and alerts, and added `L-*` labels as needed.
- [x] I have considered security and privacy, and added `I-*` labels as needed
- [x] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.

## FUD Score

_Overall, how are you feeling about these changes?_

- [x] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [x] I have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetActivityInstanceRouteTest.java [hunk 1]
```diff
@@ -155,7 +155,7 @@ private static void setupActivityAndInstance(Handle handle) throws Exception {
         txt2 = TextQuestionDef.builder(TextInputType.TEXT, "TEXT_DRUG", newTemplate())
                 .setSuggestionType(SuggestionType.DRUG)
                 .build();
-        Template txt3Tmpl = Template.html("$foo $DDP_PARTICIPANT_FIRST_NAME's favorite color?");
+        Template txt3Tmpl = Template.html("$foo $ddp.participantFirstName()'s favorite color?");
```
Comment (yufengwng) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetActivityInstanceRouteTest.java:158]: Need to match method name exactly. Can also use the other curly syntax: `${ddp.participantFirstName()}`.

File: pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetActivityInstanceRouteTest.java [hunk 2]
```diff
@@ -200,7 +200,7 @@ private static void setupActivityAndInstance(Handle handle) throws Exception {
         Template contentBody = new Template(TemplateType.HTML, null, "<p>hello body</p>");
         ContentBlockDef contentDef = new ContentBlockDef(contentTitle, contentBody);
         ContentBlockDef content2 = new ContentBlockDef(null, Template.html(
-                "<p>$DDP_PARTICIPANT_FIRST_NAME<br/>$DDP_PARTICIPANT_LAST_NAME<br/>$DDP_DASHED_DATE</p>"));
+                "<p>$ddp.participantFirstName()<br/>$ddp.participantLastName()<br/>$ddp.date(\"MM-dd-uuuu\")</p>"));
```
Comment (yufengwng) [inline pepper-apis/src/test/java/org/broadinstitute/ddp/route/GetActivityInstanceRouteTest.java:207]: We can now control how we want the date to be formatted without adding new code:
```
$ddp.date("uuuu-MM-dd")    => 2020-06-15
$ddp.date("MM-dd-uuuu")    => 06-15-2020
$ddp.date("MM/dd/uuuu")    => 06/15/2020
$ddp.date("MM/uuuu")       => 06/2020
```

**Comment by charlottedye** [comment]
This is neat!  Is this documented anywhere?  I can see it being useful to have both information on how to use the current behavior and what it is (i.e., which profile fields are currently available and how to get them, and mentioning that you can specify the format of the date) as well as a brief summary of what is involved in adding support for additional profile fields.  The first part is probably more important than the second.

**Comment by charlottedye** [comment]
Also, could you explain the snapshot piece a little bit?  Is it used for caching purposes or for something else?

**Comment by yufengwng** [comment]
@charlottedye I can write up some documentation when this proposal gets accepted/merged. Snapshotting is a feature that is toggled by a `snapshotSubstitutionsOnSubmit` flag in an activity's definition. This is so that we can save a "snapshot" of the substitution values that were used to render the activity. This is important for some activities, such as a consent form where you want the user to see exactly the same stuff they submitted.

**Comment by charlottedye** [comment]
Now this makes more sense!  It's a snapshot so that if the profile information changes, the values shown in the survey can be maintained.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/content/I18nContentRenderer.java [hunk 3]
```diff
@@ -40,6 +42,32 @@ private static Long getDefaultLanguageId(Handle handle) {
...
+        if (profile != null) {
+            if (profile.getFirstName() != null) {
+                builder.setParticipantFirstName(profile.getFirstName());
+            }
+            if (profile.getLastName() != null) {
+                builder.setParticipantLastName(profile.getLastName());
+            }
+        }
+
+        // Override with saved snapshot substitutions.
```
Comment (charlottedye) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/content/I18nContentRenderer.java]: You could probably make this comment more helpful by adding that the saved snapshot substitutions are for maintaining consistency for the user by keeping track of the value that was substituted when the user filled out the survey to show when they view the survey after they submit it when the study is configured to do so.  It's possible to piece that information together, but knowing that makes it easier to understand the code.

**Review by charlottedye — DISMISSED** [review]
Looks good to me, although I added a note about where I think you could improve a comment.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/content/I18nContentRenderer.java (hunk 3 above)
Comment (yufengwng) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/content/I18nContentRenderer.java]: done

**Review by charlottedye — APPROVED** [review]

**Pull request merged**